### PR TITLE
[1510] Handle multi-groups in form/formdescriptioneditor

### LINKED ISSUE #1510: Handle multi-groups in form/formdescriptioneditor
* [x] **I have checked that this feature has not yet been suggested by someone else.**

### Expected behavior

Handle multi-groups in form/formdescriptioneditor.

### Cuttings backs

...

### Potential side effects

...

### Describe alternatives you've considered

...

### PR DESCRIPTION
**Pull request by AxelRICHARD**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [x] Does the title of the commit contributing the ADR start with `[doc]`?
- [x] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [x] Variables have a proper type
- [x] Functions’ arguments have a proper type
- [x] Functions’ return type are specified
- Hooks are properly typed:
	- [x] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [x] `useState<STATE_TYPE>(…)`
- [x] All components have a proper typing for their props
- [x] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [x] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [x] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [x] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/adrs/081_add_support_for_multi_groups_in_form_and_formdescriptioneditor.adoc [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+Both group display modes (list and toggleable_areas) will also be handled.
+
+=== Frontend
+
+A simple div will be used to handle the new Group.
+
+=== Backend
+
+The View DSL will be modified, introducing the following concept on EClass `FormDescription`:
+* new contained multi-valued reference : `groups` with type `GroupDescription`.
```
Comment (pcdavid) [inline doc/adrs/081_add_support_for_multi_groups_in_form_and_formdescriptioneditor.adoc:28]: Needs a blank line to be formatted correctly as a list.

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/AddGroupInput.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    public AddGroupInput() {
+        // Used by jackson
+    }
+
+    public AddGroupInput(UUID id, String editingContextId, String representationId, int index) {
+        this.id = Objects.requireNonNull(id);
+        this.editingContextId = Objects.requireNonNull(editingContextId);
+        this.representationId = Objects.requireNonNull(representationId);
+        this.index = Objects.requireNonNull(index);
```
Comment (pcdavid) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/AddGroupInput.java]: Not needed for an `int`.

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/MoveGroupInput.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    public MoveGroupInput() {
+        // Used by jackson
+    }
+
+    public MoveGroupInput(UUID id, String editingContextId, String representationId, String groupId, int index) {
+        this.id = Objects.requireNonNull(id);
+        this.editingContextId = Objects.requireNonNull(editingContextId);
+        this.representationId = Objects.requireNonNull(representationId);
+        this.groupId = Objects.requireNonNull(groupId);
+        this.index = Objects.requireNonNull(index);
```
Comment (pcdavid) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/MoveGroupInput.java]: Not needed for an `int`.

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/form/ViewFormDescriptionConverter.java [hunk 4]
```diff
@@ -123,6 +112,32 @@ public IRepresentationDescription convert(RepresentationDescription representati
...
+                .map(dispatcher::doSwitch)
+                .filter(ButtonDescription.class::isInstance)
+                .map(ButtonDescription.class::cast)
+                .collect(Collectors.toList());
+
+        Function<VariableManager, List<?>> semanticElementsProvider = variableManager -> variableManager.get(VariableManager.SELF, Object.class).stream().collect(Collectors.toList());
+
+        return GroupDescription.newGroupDescription(Optional.ofNullable(viewGroupDescription.getName()).orElse("")) //$NON-NLS-1$
+                .idProvider(new GetOrCreateRandomIdProvider())
+                .labelProvider(variableManager -> Optional.ofNullable(viewGroupDescription.getLabelExpression()).orElse("")) //$NON-NLS-1$
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/form/ViewFormDescriptionConverter.java]: The expression is not evaluated:

File: packages/view/backend/sirius-components-view/src/main/resources/model/view.ecore [hunk 5]
```diff
@@ -236,11 +236,23 @@
     <eLiterals name="UNSYNCHRONIZED" value="1"/>
   </eClassifiers>
   <eClassifiers xsi:type="ecore:EClass" name="FormDescription" eSuperTypes="#//RepresentationDescription">
+    <eStructuralFeatures xsi:type="ecore:EReference" name="groups" upperBound="-1"
+        eType="#//GroupDescription" containment="true"/>
+  </eClassifiers>
+  <eClassifiers xsi:type="ecore:EClass" name="GroupDescription">
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view/src/main/resources/model/view.ecore:242]: Why no `semanticCandidatesExpression`? `org.eclipse.sirius.components.forms.description.GroupDescription` and `GroupComponent` support them.

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveGroupEventHandler.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        if (optionalSelf.isPresent()) {
+            Object container = optionalSelf.get();
+            if (container instanceof FormDescription) {
+                var objectToMove = this.objectService.getObject(editingContext, groupId);
+                if (objectToMove.filter(GroupDescription.class::isInstance).isPresent()) {
+                    GroupDescription groupToMove = (GroupDescription) objectToMove.get();
+                    if (container.equals(groupToMove.eContainer())) {
+                        ((FormDescription) container).getGroups().move(index, groupToMove);
+                    } else {
+                        ((FormDescription) container).getGroups().add(index, groupToMove);
```
Comment (pcdavid) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveGroupEventHandler.java]: Check for index validity, or catch `IndexOutOfBoundsException` and return failure on invalid index (same as in `MoveToolbarActionEventHandler`).

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: doc/adrs/081_add_support_for_multi_groups_in_form_and_formdescriptioneditor.adoc (hunk 1 above)
Comment (AxelRICHARD) [reply doc/adrs/081_add_support_for_multi_groups_in_form_and_formdescriptioneditor.adoc:28]: Done

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/AddGroupInput.java (hunk 2 above)
Comment (AxelRICHARD) [reply packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/AddGroupInput.java]: Done

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/MoveGroupInput.java (hunk 3 above)
Comment (AxelRICHARD) [reply packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/dto/MoveGroupInput.java]: Done

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/form/ViewFormDescriptionConverter.java (hunk 4 above)
Comment (AxelRICHARD) [reply packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/form/ViewFormDescriptionConverter.java]: Done

File: packages/view/backend/sirius-components-view/src/main/resources/model/view.ecore (hunk 5 above)
Comment (AxelRICHARD) [reply packages/view/backend/sirius-components-view/src/main/resources/model/view.ecore:242]: Done

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveGroupEventHandler.java (hunk 6 above)
Comment (AxelRICHARD) [reply packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveGroupEventHandler.java]: Done

File: packages/view/backend/sirius-components-view/src/main/resources/model/view.ecore [hunk 7]
```diff
@@ -236,11 +236,25 @@
...
+    <eStructuralFeatures xsi:type="ecore:EAttribute" name="name" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
+    <eStructuralFeatures xsi:type="ecore:EAttribute" name="labelExpression" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString"/>
+    <eStructuralFeatures xsi:type="ecore:EAttribute" name="displayMode" lowerBound="1"
+        eType="#//GroupDisplayMode" defaultValueLiteral="LIST"/>
+    <eStructuralFeatures xsi:type="ecore:EAttribute" name="semanticCandidatesExpression"
+        eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString" defaultValueLiteral="aql:self.eContents()"/>
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view/src/main/resources/model/view.ecore]: The `aql:self.eContents()` default is what causes test failures. I think by default we should put `aql:self` to keep the previous behavior and let users opt-in to the possibility of multiple group instances.

**Review by pcdavid — CHANGES_REQUESTED** [review]

**Comment by pcdavid** [comment]
When actually using the new capability to instanciate multiple groups from the same description, I get React errors about duplicate keys in the front.

**Comment by pcdavid** [comment]
`ToolbarButtons` do not work: when clicked on, they send a `pushButton` mutation, but `PushButtonEventHandler` only supports actual `Button`s, which are now separate from `ToolbarAction`:

```java
            // @formatter:off
            IStatus status = this.formQueryService.findWidget(form, input.getButtonId())
                    .filter(Button.class::isInstance)
                    .map(Button.class::cast)
                    .map(Button::getPushButtonHandler)
                    .map(Supplier::get)
                    .orElse(new Failure("")); //$NON-NLS-1$
            // @formatter:on
```

From `.filter(Button.class::isInstance)`, the stream is empty.

It's not related to this particular PR, it was probably introduced earlier but not seen. I can open a separate ticket if you prefer.

**Comment by pcdavid** [comment]
Using this fixes the issue:
```java
            // @formatter:off
            Optional<AbstractWidget> optionalWidget = this.formQueryService.findWidget(form, input.getButtonId());
            var handler = optionalWidget.filter(Button.class::isInstance)
                                        .map(Button.class::cast)
                                        .map(Button::getPushButtonHandler);
            if (handler.isEmpty()) {
                handler = optionalWidget.filter(ToolbarAction.class::isInstance)
                                        .map(ToolbarAction.class::cast)
                                        .map(ToolbarAction::getPushButtonHandler);
            }
            IStatus status = handler.map(Supplier::get).orElse(new Failure("")); //$NON-NLS-1$
            // @formatter:on
```

File: packages/view/backend/sirius-components-view/src/main/resources/model/view.ecore (hunk 7 above)
Comment (AxelRICHARD) [reply packages/view/backend/sirius-components-view/src/main/resources/model/view.ecore]: Done

**Comment by AxelRICHARD** [comment]
duplicate keys problem and PushButtonEventHandler also fixed

**Review by pcdavid — APPROVED** [review]

**Pull request merged**